File: src/Control/Monad/ST/Internal.php
<?php

$map_ = function($f, $a = null) use (&$map_) {
    if (\func_num_args() < 2) {
        $__args = \func_get_args();
        return function(...$more) use ($__args, &$map_) {

            return $map_(...\array_merge($__args, $more));
        };
    }
    return function() use(&$f, &$a) { return $f($a()); };
};
$bind_ = function($a, $f = null) use (&$bind_) {
    if (\func_num_args() < 2) {
        $__args = \func_get_args();
        return function(...$more) use ($__args, &$bind_) {

            return $bind_(...\array_merge($__args, $more));
        };
    }
    return function() use(&$a, &$f) { return $f($a())(); };
};
$pure_ = function($a) use (&$pure_) { return function() use(&$a) { return $a; }; };
$new = function($val) use (&$new) { return function() use(&$val) { return (object)['value' => $val]; }; };
$read = function($ref) use (&$read) { return function() use(&$ref) { return $ref->value; }; };
$modifyImpl = function($f, $ref = null) use (&$modifyImpl) {
    if (\func_num_args() < 2) {
        $__args = \func_get_args();
        return function(...$more) use ($__args, &$modifyImpl) {

            return $modifyImpl(...\array_merge($__args, $more));
        };
    }
    return function() use(&$f, &$ref) { $t = $f($ref->value); $ref->value = $t->state; return $t->value; };
};
$write = function($val, $ref = null) use (&$write) {
    if (\func_num_args() < 2) {
        $__args = \func_get_args();
        return function(...$more) use ($__args, &$write) {

            return $write(...\array_merge($__args, $more));
        };
    }
    return function() use(&$val, &$ref) { $ref->value = $val; return $val; };
};
$run = function($f) use (&$run) { return $f(); };
$while = function($f, $a = null) use (&$while) {
    if (\func_num_args() < 2) {
        $__args = \func_get_args();
        return function(...$more) use ($__args, &$while) {

            return $while(...\array_merge($__args, $more));
        };
    }
    return function() use(&$f, &$a) { while ($f()) { $a(); } return null; };
};
$for = function($lo, $hi = null, $f = null) use (&$for) {
    if (\func_num_args() < 3) {
        $__args = \func_get_args();
        return function(...$more) use ($__args, &$for) {

            return $for(...\array_merge($__args, $more));
        };
    }
    return function() use(&$lo, &$hi, &$f) { for ($i = $lo; $i < $hi; $i++) { $f($i)(); } return null; };
};
$foreach = function($as, $f = null) use (&$foreach) {
    if (\func_num_args() < 2) {
        $__args = \func_get_args();
        return function(...$more) use ($__args, &$foreach) {

            return $foreach(...\array_merge($__args, $more));
        };
    }
    return function() use(&$as, &$f) { foreach ($as as $a) { $f($a)(); } return null; };
};

$exports['map_'] = $map_;
$exports['bind_'] = $bind_;
$exports['pure_'] = $pure_;
$exports['new'] = $new;
$exports['read'] = $read;
$exports['modifyImpl'] = $modifyImpl;
$exports['write'] = $write;
$exports['run'] = $run;
$exports['while'] = $while;
$exports['for'] = $for;
$exports['foreach'] = $foreach;
return $exports;

File: src/Control/Monad/ST/Internal.fs
let map_ = 
    fun (f: obj) -> fun (a: obj) ->
        let a' = a :?> (unit -> obj)
        (fun () -> Sharpurs_Prelude.sharpurs_apply f (a' ())) :> obj

let bind_ = 
    fun (a: obj) -> fun (f: obj) ->
        let a' = a :?> (unit -> obj)
        (fun () -> 
            let res = Sharpurs_Prelude.sharpurs_apply f (a' ()) :?> (unit -> obj)
            res ()) :> obj

let pure_ = 
    fun (a: obj) -> 
        (fun () -> a) :> obj

let run = 
    fun (f: obj) -> 
        let f' = f :?> (unit -> obj)
        f' ()

let ``while`` = 
    fun (cond: obj) -> fun (a: obj) ->
        let cond' = cond :?> (unit -> obj)
        let a' = a :?> (unit -> obj)
        (fun () ->
            while (unbox<bool> (cond' ())) do
                a' () |> ignore
            null :> obj) :> obj

let ``new`` = 
    fun (val': obj) -> 
        (fun () -> ref val' :> obj) :> obj

let read = 
    fun (r: obj) -> 
        let r' = r :?> obj ref
        (fun () -> r'.Value) :> obj

let modifyImpl = 
    fun (f: obj) -> fun (r: obj) ->
        let r' = r :?> obj ref
        (fun () ->
            let res = Sharpurs_Prelude.sharpurs_apply f (r'.Value) :?> Map<string, obj>
            r'.Value <- Map.find "state" res
            Map.find "value" res) :> obj

let write = 
    fun (a: obj) -> fun (r: obj) ->
        let r' = r :?> obj ref
        (fun () ->
            r'.Value <- a
            a) :> obj

let ``for`` =
    fun (lo: obj) -> fun (hi: obj) -> fun (f: obj) ->
        let lo' = lo :?> int
        let hi' = hi :?> int
        (fun () ->
            for i = lo' to hi' - 1 do
                let step = Sharpurs_Prelude.sharpurs_apply f (box i) :?> (unit -> obj)
                step () |> ignore
            null :> obj) :> obj

let foreach =
    fun (xs: obj) -> fun (f: obj) ->
        let arr = xs :?> obj[]
        (fun () ->
            for x in arr do
                let step = Sharpurs_Prelude.sharpurs_apply f x :?> (unit -> obj)
                step () |> ignore
            null :> obj) :> obj
